### PR TITLE
[MRG+1] Correct theta values when drawing a non-circular arc

### LINKED ISSUE #8046: Arc patch with starting and ending angle
### Bug report

**Bug summary**

- Arc patch does not behave as expected when drawing elliptical arcs between two angles specified by theta1 and theta2.

**Code for reproduction**
```python
import matplotlib.pyplot as plt
from matplotlib.patches import Arc

# Ellipse parameters
R = 1.0
x, y = 0.0, 0.0
a, b = 2.0*R, R

# Figure setup
fig_width, fig_height = 3.30, 3.30
fig = plt.figure(figsize=(fig_width, fig_height), frameon=False)
ax = fig.add_axes([0.0, 0.0, 1.0, 1.0], aspect='equal')
ax.set_axis_off()
ax.set_xlim(-R*1.05, R*1.05)
ax.set_ylim(-R*1.05, R*1.05)
# Axes
ax.axhline(0.0)
ax.axvline(0.0)
# 45 degree line
ax.plot([0.0, 1.0], [0.0, 1.0], 'k--')
# Arcs
ax.add_patch(Arc((x, y), a, b, 
                 theta1=0.0, theta2=360.0, edgecolor='k'))
ax.add_patch(Arc((x, y), a, b,
                 theta1=0.0, theta2=45.0, edgecolor='r', lw=1.5))
fig.savefig('Arc_patch_bug.png')
```

**Actual outcome**

arc_patch_bug

**Expected outcome**

- Red elliptical arc between the x-axis (0 degrees) and the dashed line (45 degrees), following the black ellipse.

**Matplotlib version**

- Matplotlib 2.0.0, Python 2.7.13, OSX
- Installed with MacPorts

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #8046. I'm not sure if this is the right place to do the correction, so a second opinion would be good. I've checked this works with a full range of angles.

### CODE REVIEW

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -982,6 +982,23 @@ def test_canonical():
     ax.plot([1, 2, 3])
 
 
+@image_comparison(baseline_images=['arc_angles'])
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py]: add `remove_text=True` and `style='default'` to the decorator to get the new defaults + drop all of the text (which makes the test more stable).

**Comment by tacaswell** [comment]
This will need an entry in whats new or api changes.  The paranoid case here is that users have worked around this bug and we are now silently changing things under them so it needs to be documented.

**Comment by tacaswell** [comment]
This looks like the correct place to do this to me.  `Ellipse` is implemented by holding transforms to re-scale a circular path so pushing this down into `Path.arc` is not right and `Ellipse` is for full coverage so pushing this into the base-class is not right either.

**Comment by dstansby** [comment]
Does anyone know why the documentation build is failing with:

```
/Users/dstansby/matplotlib/doc/examples/units/ellipse_with_units.rst:8: WARNING: Exception occurred in plotting ellipse_with_units
 from /Users/dstansby/matplotlib/doc/mpl_examples/units/ellipse_with_units.py:
Traceback (most recent call last):
  File "/Users/dstansby/matplotlib/lib/matplotlib/sphinxext/plot_directive.py", line 525, in run_code
    six.exec_(code, ns)
  File "<string>", line 55, in <module>
  File "/Users/dstansby/matplotlib/lib/matplotlib/patches.py", line 1565, in __init__
    theta1 = theta_stretch(theta1, width / height)
TypeError: unsupported operand type(s) for /: 'TaggedValue' and 'TaggedValue'
```

**Comment by afvincent** [comment]
As far as I understand, it seems like “width” and “height” are `cm` instances from this [script](http://matplotlib.org/examples/units/basic_units.html) and that the class used under the hood (`TaggedValue`) does not implement the division operation.

**Comment by dstansby** [comment]
Would anyone object to me just getting rid of http://matplotlib.org/examples/units/ellipse_with_units.html ? It doesn't seem to be showing any unique capabilities of units, and the 'comparison' just looks the same to me.

**Comment by tacaswell** [comment]
Yes and yes (although, it seems like the problem here is the exact unit implementation, not generally).  

The unit support is an under-understood, but critical feature of mpl (JPL relies on it).

**Comment by dstansby** [comment]
As far as I can tell:

- The correct way to convert from units to a number is call `self.convert_xunits(width)`
- However this is just returning `width` unchanged as units
- This is because there is no axis when `__init__` is called on `Arc`, so the units are returned unchanged

so all this means there's no way of converting units within the `__init__` bit of `Arc`, so the conversion needs to happen somewhere else.

**Comment by dstansby** [comment]
That should fix it, I've moved the new calculation into the `draw()` method.

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -982,6 +982,27 @@ def test_canonical():
...
+    w = 2
+    h = 1
+    centre = (0.2, 0.5)
+
+    fig, axs = plt.subplots(3, 3)
+    for i, ax in enumerate(np.ravel(axs)):
+        theta2 = i * 360 / 9
+        theta1 = theta2 - 45
+        ax.add_patch(patches.Arc(centre, w, h, theta1=theta1, theta2=theta2))
+        # Straight line should intersect end of arc
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Could you also add the start of the arc as well? Start -> origin -> end should work.

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -982,6 +982,27 @@ def test_canonical():
...
+                  style='default')
+def test_arc_angles():
+    from matplotlib import patches
+    # Ellipse parameters
+    w = 2
+    h = 1
+    centre = (0.2, 0.5)
+
+    fig, axs = plt.subplots(3, 3)
+    for i, ax in enumerate(np.ravel(axs)):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: `enumerate(axs.flat)`

File: lib/matplotlib/tests/test_axes.py [hunk 4]
```diff
@@ -982,6 +982,27 @@ def test_canonical():
...
+        theta2 = i * 360 / 9
+        theta1 = theta2 - 45
+        ax.add_patch(patches.Arc(centre, w, h, theta1=theta1, theta2=theta2))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Can you also add a full `Ellipse` patch with the same parameters but maybe half linewidth or alpha just to show that the arc is in a consistent place?

**Comment by dstansby** [comment]
Thanks for the extra ideas for the test @QuLogic - once this has been reviewed I can squash everything down to a single commit.

**Comment by QuLogic** [comment]
Is there anything backend-specific here? Do we need the SVG/PDF files?

**Comment by QuLogic** [comment]
Yes, I think we can drop the other ones and you can squash this.

**Comment by dstansby** [comment]
Great, that should be all nicely squashed up now.

File: doc/api/api_changes/2017-02-10-DS_elliptical_arc_angle.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Elliptical arcs now drawn between correct angles
+````````````````````````````````````````````````
+
+The `matplotlib.patches.Arc` patch is now correctly drawn between the given
+angles.
+
+Previously a circular are was drawn and then stretched into an ellipse,
```
Comment (QuLogic) [inline doc/api/api_changes/2017-02-10-DS_elliptical_arc_angle.rst]: are -> arc

Comment (dstansby) [inline doc/api/api_changes/2017-02-10-DS_elliptical_arc_angle.rst]: woops...

**Pull request merged**